### PR TITLE
Add index for mined blocks

### PR DESCRIPTION
**Pull request by deekerno**
## Summary
This adds an indexer for mined blocks. Adjustments to the `miners:mined` CLI command are coming in a separate PR.

## Testing Plan
Added tests and ensured that node runs as expected.

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and
what additional work is required, if any.

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: ironfish/src/node.ts [hunk 1]
```diff
@@ -284,6 +304,7 @@ export class IronfishNode {
     }
 
     this.telemetry.submitNodeStarted()
+    this.minedBlocksIndexer.start()
```
Comment (NullSoldier) [inline ironfish/src/node.ts:307]: The telemetry is wrong now based on this ordering right?

File: ironfish/src/indexer/indexer.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  constructor({ files, location }: { files: FileSystem; location: string }) {
+    this.files = files
+    this.location = location
+    this.database = createDB({ location })
+  }
+
+  async openDB(options: { upgrade?: boolean } = { upgrade: true }): Promise<void> {
+    await this.files.mkdir(this.location, { recursive: true })
+    await this.database.open()
+    if (options.upgrade) {
```
Comment (NullSoldier) [inline ironfish/src/indexer/indexer.ts:24]: nit: extra space here? :)
```suggestion

    if (options.upgrade) {
```

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+      valueEncoding: new JsonEncoding(),
+    })
+
+    this.chainProcessor.onAdd.on(async (header) => {
+      const block = await chain.getBlock(header)
+      Assert.isNotNull(block)
+
+      const account = this.accounts.listAccounts().find((a) => isBlockMine(block, a))
+      if (account) {
+        const main = await this.chain.isHeadChain(header)
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:88]: this should always be true here

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        const minersFee = this.chain.strategy.miningReward(header.sequence)
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:89]: You don't need to recalculate this, `header.minersFee`

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    })
+
+    this.chainProcessor.onRemove.on(async (header) => {
+      if (await this.minedBlocks.has(header.hash.toString('hex'))) {
+        const block = await chain.getBlock(header)
+        Assert.isNotNull(block)
+
+        const account = this.accounts.listAccounts().find((a) => isBlockMine(block, a))
+        if (account) {
+          const main = await this.chain.isHeadChain(header)
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:104]: this should always be false

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+          const minersFee = this.chain.strategy.miningReward(header.sequence)
+
+          await this.put(header, account.name, main, minersFee)
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:107]: this is already stored, you just need to edit the existing one right?

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        }
+      }
+      await this.updateHeadHash(header.previousBlockHash)
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:110]: what's the implication of not doing all this inside of a transaction?

File: ironfish/src/indexer/indexer.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { FileSystem } from '../fileSystems'
+import { IDatabase } from '../storage/database'
+import { createDB } from '../storage/utils'
+
+const DATABASE_VERSION = 1
+
+export abstract class Indexer {
```
Comment (NullSoldier) [inline ironfish/src/indexer/indexer.ts:10]: This is called Indexer, but what logic here is related to indexing? Isn't that all in MinedBlocksIndexer? Most of the abstraction isn't in the indexer.

File: ironfish/src/indexer/minedBlocksIndexer.ts (hunk 3 above)
Comment (deekerno) [reply ironfish/src/indexer/minedBlocksIndexer.ts:88]: Good catch, thanks!

File: ironfish/src/indexer/minedBlocksIndexer.ts (hunk 5 above)
Comment (deekerno) [reply ironfish/src/indexer/minedBlocksIndexer.ts:104]: Good catch, thanks!

File: ironfish/src/node.ts (hunk 1 above)
Comment (deekerno) [reply ironfish/src/node.ts:307]: Good catch, thanks!

File: ironfish/src/indexer/minedBlocksIndexer.ts (hunk 7 above)
Comment (deekerno) [reply ironfish/src/indexer/minedBlocksIndexer.ts:105]: Changes would not be atomic, which could lead to non-linearizability. Adding the transaction.

File: ironfish/src/account/accounts.ts [hunk 9]
```diff
@@ -32,6 +32,8 @@ export class Accounts {
     [account: Account | null, oldAccount: Account | null]
   >()
 
+  readonly onAccountCreated = new Event<[account: Account, imported: boolean]>()
```
Comment (NullSoldier) [inline ironfish/src/account/accounts.ts:35]: if something is imported, we're not creating an account, we're importing it. I think the bool parameter and the name of this are mutually exclusive.

File: ironfish/src/account/accounts.ts [hunk 10]
```diff
@@ -957,6 +963,8 @@ export class Accounts {
 
     this.accounts.delete(name)
     await this.db.removeAccount(name)
+
+    await this.onAccountRemoved.emitAsync(name)
```
Comment (NullSoldier) [inline ironfish/src/account/accounts.ts:967]: Do we need async properties for these events? Should the accounts need to wait for the receiver to finish the event callback?

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+    })
+
+    this.chainProcessor = new ChainProcessor({ logger, chain, head: null })
+
+    this.chainProcessor.onAdd.on(async (header) => {
+      const block = await this.chain.getBlock(header)
+      Assert.isNotNull(block)
+
+      const account = this.accounts.listAccounts().find((a) => isBlockMine(block, a))
+      await this.database.transaction(async (tx) => {
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:98]: I think there's a few things here worth thinking through

1. It's possible to start a transaction here that does nothing in the current code
2. There's a data consistency issue with this transaction.

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+              hash: header.hash.toString('hex'),
+              sequence: header.sequence,
+              account: account.name,
+              minersFee: Number(header.minersFee),
+            },
+            tx,
+          )
+          const hashes: string[] = (await this.accountToBlockHashes.get(account.name, tx)) || []
+          hashes.concat(block.header.hash.toString('hex'))
+          await this.accountToBlockHashes.put(account.name, hashes, tx)
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:113]: this will store the entire first table into a value of row in the second table.

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+      Assert.isNotNull(block)
+
+      const account = this.accounts.listAccounts().find((a) => isBlockMine(block, a))
+      await this.database.transaction(async (tx) => {
+        if (account) {
+          await this.minedBlocks.put(
+            header.hash.toString('hex'),
+            {
+              main: true,
+              hash: header.hash.toString('hex'),
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:104]: We're double storing the hash here.

File: ironfish/src/indexer/minedBlocksIndexer.ts (hunk 7 above)
Comment (NullSoldier) [reply ironfish/src/indexer/minedBlocksIndexer.ts:135]: I'm unresolving, because a transaction wasn't added.

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+    })
+
+    this.accounts.onAccountCreated.on((_account, imported) => {
+      if (imported) {
+        this.chainProcessor.hash = null
+      }
+    })
+
+    this.accounts.onAccountRemoved.on(async (accountName) => {
+      await this.database.transaction(async (tx) => {
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:145]: Is this just used as a lock to avoid doing this while the chain processor is running?

File: ironfish/src/indexer/minedBlocksIndexer.ts [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+    const minedBlocks = (await this.minedBlocks.getAllValues()).sort(
+      (a, b) => a.sequence - b.sequence,
+    )
+
+    if (includeForks) {
+      return minedBlocks
+    }
+
+    return minedBlocks.filter((block) => block.main === true)
+  }
```
Comment (NullSoldier) [inline ironfish/src/indexer/minedBlocksIndexer.ts:267]: How do you filter on the sequence?

Also, is it possible for us to get away with just yielding blocks instead of accumulating and sorting all the blocks?

File: ironfish/src/account/accounts.ts (hunk 9 above)
Comment (deekerno) [reply ironfish/src/account/accounts.ts:35]: Sure, I can replace this with `onAccountImported` since we don't really care about account creation in this indexer.

File: ironfish/src/indexer/minedBlocksIndexer.ts (hunk 15 above)
Comment (deekerno) [reply ironfish/src/indexer/minedBlocksIndexer.ts:267]: My intention was to filter in the RPC route and have this strictly return the desired blocks, but there's really no reason that it couldn't be done here by passing in options instead. Regarding your second question, since we're using hashes as the key for the store, `.getAllValues()` ends up being lexicographically sorted by hash. The current output of `miners:mined` is sorted by sequence.

File: ironfish/src/indexer/minedBlocksIndexer.ts (hunk 14 above)
Comment (deekerno) [reply ironfish/src/indexer/minedBlocksIndexer.ts:145]: Essentially, yes. I'm ensuring that it's an atomic change.

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+      .sort((a, b) => a.sequence - b.sequence)
+
+    if (!scanForks) {
+      minedBlocks = minedBlocks.filter((block) => block.main === true)
+    }
+
+    for (const block of minedBlocks) {
+      yield block
+    }
+  }
```
Comment (NullSoldier) [inline ironfish/src/indexers/minedBlocksIndexer.ts:292]: We should make this actually able to iterate the results lazily, but to do that we need to avoid accumulating all the results. We can do that with an index by sequence.

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 17]
```diff
@@ -215,16 +199,36 @@ export class MinedBlocksIndexer {
...
   }
 
-  start(): void {
+  async start(): Promise<void> {
     if (this.isStarted) {
       return
     }
     this.isStarted = true
 
+    if (this.removalInProgress) {
```
Comment (NullSoldier) [inline ironfish/src/indexers/minedBlocksIndexer.ts:225]: If you just remove this, I think it'll work the same. It's just an extra variable that's derived from `accountToRemove` anyway.

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 18]
```diff
@@ -215,16 +199,36 @@ export class MinedBlocksIndexer {
...
+      const accountName = await this.meta.get('accountToRemove')
+      if (accountName) {
+        await this.removeMinedBlocks(accountName)
+      }
+    }
```
Comment (NullSoldier) [inline ironfish/src/indexers/minedBlocksIndexer.ts:230]: Code architecture wise, all this should be inside the event loop to avoid updating the head, and handling an account removal.

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 19]
```diff
@@ -166,14 +157,7 @@ export class MinedBlocksIndexer {
...
     this.accounts.onAccountRemoved.on(async (account) => {
-      await this.database.transaction(async (tx) => {
-        const hashes = await this.accountToBlockHashes.get(account.name, tx)
-        if (hashes) {
-          for (const hash of hashes) {
-            await this.minedBlocks.del(hash, tx)
-          }
-        }
-      })
+      await this.removeMinedBlocks(account.name)
```
Comment (NullSoldier) [inline ironfish/src/indexers/minedBlocksIndexer.ts:160]: I think all this should be doing is queueing the "job" for removal, and letting the "scheduler" pick up that job. It ensures that no two things are running that could be conflicting. Like a prune job and updating the index.

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 20]
```diff
@@ -270,6 +274,37 @@ export class MinedBlocksIndexer {
     await this.meta.put('headHash', hashString, tx)
   }
 
+  async removeMinedBlocks(accountName: string): Promise<void> {
+    await this.meta.put('accountToRemove', accountName)
+
+    const hashes: Buffer[] = []
+    const sequences: number[] = []
+
+    await this.database.transaction(async (tx) => {
```
Comment (NullSoldier) [inline ironfish/src/indexers/minedBlocksIndexer.ts:283]: We probably only need a transaction around the operations and not loading the entire database for pruning. Transactions cache all read state, so this will keep everything loaded in memory because of how this works. It also holds a really long lock.

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 21]
```diff
@@ -0,0 +1,344 @@
...
+    }
+  }
+
+  async updateHeadHash(headHash: Buffer | null, tx?: IDatabaseTransaction): Promise<void> {
+    const hashString = headHash && headHash.toString('hex')
+    await this.meta.put('headHash', hashString, tx)
+  }
+
+  async removeMinedBlocks(accountName: string): Promise<void> {
+    await this.meta.put('accountToRemove', accountName)
```
Comment (NullSoldier) [inline ironfish/src/indexers/minedBlocksIndexer.ts:271]: I think this is left over

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 22]
```diff
@@ -272,7 +272,9 @@ export class MinedBlocksIndexer {
     for await (const [hash, block] of iterator) {
       if (block.account === accountName) {
         await this.database.transaction(async (tx) => {
-          await this.sequenceToHashes.del(block.sequence, tx)
+          let hashes = (await this.sequenceToHashes.get(block.sequence)) ?? []
```
Comment (NullSoldier) [inline ironfish/src/indexers/minedBlocksIndexer.ts:275]: It looks like you're using null coalescence to default this to [] because it can be undefined, but how can it be undefined?

Comment (deekerno) [reply ironfish/src/indexers/minedBlocksIndexer.ts:275]: That's a fair point. After looking through the code paths, it shouldn't ever be undefined if a block was added correctly. I think doing `Assert.isNotUndefined()` would be a good replacement.

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 23]
```diff
@@ -0,0 +1,369 @@
...
+    this.isStarted = false
+    this.rescan = false
+
+    this.meta = this.database.addStore<{
+      key: keyof MinedBlocksDBMeta
+      value: MinedBlocksDBMeta[keyof MinedBlocksDBMeta]
+    }>({
+      name: 'meta',
+      keyEncoding: new StringEncoding<keyof MinedBlocksDBMeta>(),
+      valueEncoding: new JsonEncoding(),
```
Comment (rohanjadvani) [inline ironfish/src/indexers/minedBlocksIndexer.ts:92]: We now have binary encodings for all our stores in the node. Can you add encoders for each of these stores? You can look at the [PR](https://github.com/iron-fish/ironfish/pull/1256) here for examples. [Here](https://github.com/iron-fish/ironfish/blob/master/ironfish/src/account/database/accounts.ts) is how accounts are stored.

It might be useful to do these as separate PRs into this branch.

File: ironfish/src/indexers/database/accountsToRemove.ts [hunk 24]
```diff
@@ -0,0 +1,44 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (rohanjadvani) [inline ironfish/src/indexers/database/accountsToRemove.ts:1]: Can you add unit tests for each of these encoders?

File: ironfish/src/indexers/database/meta.ts [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+  headHash: string | null
+}
+
+export type MetaValue = MinedBlocksDBMeta[keyof MinedBlocksDBMeta]
+
+export class MetaValueEncoding implements IDatabaseEncoding<MetaValue> {
+  serialize(value: MetaValue): Buffer {
+    const bw = bufio.write(this.getSize(value))
+    if (value) {
+      bw.writeVarString(value, 'utf8')
```
Comment (rohanjadvani) [inline ironfish/src/indexers/database/meta.ts:17]: Should this use `writeHash`?

File: ironfish/src/indexers/database/sequenceToHashes.ts [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+
+    while (reader.left()) {
+      hashes.push(reader.readHash())
+    }
+
+    return { hashes }
+  }
+
+  getSize(value: SequenceToHashesValue): number {
+    let size = 0
```
Comment (rohanjadvani) [inline ironfish/src/indexers/database/sequenceToHashes.ts:36]:
```suggestion
    return 32 * value.hashes.length
```

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 27]
```diff
@@ -0,0 +1,395 @@
...
+} from './database/sequenceToHashes'
+
+const DATABASE_VERSION = 1
+const REMOVAL_KEY = 'accountsToRemove'
+
+const getMinedBlocksDBMetaDefaults = (): MinedBlocksDBMeta => ({
+  headHash: null,
+})
+
+type MinedBlocksDBMeta = {
```
Comment (rohanjadvani) [inline ironfish/src/indexers/minedBlocksIndexer.ts:39]: This seems like a duplicate type now

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 28]
```diff
@@ -0,0 +1,395 @@
...
+      await this.accountsToRemove.put(REMOVAL_KEY, { accounts: accountsToRemove })
+    }
+
+    this.logger.debug(`Finished removing mined blocks for account ${accountName}`)
+  }
+
+  async getMinedBlock(
+    blockHash: Buffer,
+  ): Promise<
+    | { main: boolean; sequence: number; account: string; minersFee: number; hash: string }
```
Comment (rohanjadvani) [inline ironfish/src/indexers/minedBlocksIndexer.ts:342]: Does this work?

```suggestion
    MinedBlockValue | undefined
```

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 29]
```diff
@@ -0,0 +1,395 @@
...
+  async *getMinedBlocks({
+    scanForks,
+    start,
+    stop,
+  }: {
+    scanForks?: boolean
+    start?: number
+    stop?: number
+  }): AsyncGenerator<
+    { main: boolean; sequence: number; account: string; minersFee: number; hash: string },
```
Comment (rohanjadvani) [inline ironfish/src/indexers/minedBlocksIndexer.ts:359]:
```suggestion
    MinedBlockValue
```

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 30]
```diff
@@ -0,0 +1,395 @@
...
+      name: 'blocks',
+      keyEncoding: new BufferEncoding(),
+      valueEncoding: new MinedBlockValueEncoding(),
+    })
+
+    this.sequenceToHashes = this.database.addStore<{
+      key: number
+      value: SequenceToHashesValue
+    }>({
+      name: 'seqToHash',
```
Comment (rohanjadvani) [inline ironfish/src/indexers/minedBlocksIndexer.ts:106]:
```suggestion
      name: 'sequenceToHashes',
```

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 31]
```diff
@@ -0,0 +1,395 @@
...
+      key: keyof MinedBlocksDBMeta
+      value: MinedBlocksDBMeta[keyof MinedBlocksDBMeta]
+    }>({
+      name: 'meta',
+      keyEncoding: new StringEncoding<keyof MinedBlocksDBMeta>(),
+      valueEncoding: new MetaValueEncoding(),
+    })
+
+    this.minedBlocks = this.database.addStore<{ key: Buffer; value: MinedBlockValue }>({
+      name: 'blocks',
```
Comment (rohanjadvani) [inline ironfish/src/indexers/minedBlocksIndexer.ts:97]:
```suggestion
      name: 'minedBlocks',
```

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 32]
```diff
@@ -0,0 +1,395 @@
...
+      name: 'seqToHash',
+      keyEncoding: new U32Encoding(),
+      valueEncoding: new SequenceToHashesValueEncoding(),
+    })
+
+    this.accountsToRemove = this.database.addStore<{
+      key: string
+      value: AccountsToRemoveValue
+    }>({
+      name: 'accsToRemove',
```
Comment (rohanjadvani) [inline ironfish/src/indexers/minedBlocksIndexer.ts:115]:
```suggestion
      name: 'accountsToRemove',
```

**Review by rohanjadvani — COMMENTED** [review]
Let's be explicit with the names. We don't need to shorten these.

**Comment by deekerno** [comment]
> Let's be explicit with the names. We don't need to shorten these.

@NullSoldier recommended keeping the names short due to how LevelDB represents rows under the hood. Happy to change it either way.

File: ironfish/src/indexers/database/meta.ts (hunk 25 above)
Comment (deekerno) [reply ironfish/src/indexers/database/meta.ts:17]: It should!

File: ironfish/src/indexers/database/accountsToRemove.test.ts [hunk 33]
```diff
@@ -0,0 +1,34 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { AccountsToRemoveValue, AccountsToRemoveValueEncoding } from './accountsToRemove'
+
+describe('AccountsToRemoveValueEncoding', () => {
+  describe('with an empty list', () => {
```
[the remark this answers is not in the document]
Comment (deekerno) [reply ironfish/src/indexers/database/accountsToRemove.test.ts:7]: 👍🏽

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 34]
```diff
@@ -0,0 +1,391 @@
...
+    const { hashChanged } = await this.chainProcessor.update()
+
+    if (hashChanged) {
+      this.logger.debug(
+        `Updated MinedBlocksIndexer Head: ${String(this.chainProcessor.hash?.toString('hex'))}`,
+      )
+    }
+  }
+
+  async getHashesAtSequence(sequence: number, tx?: IDatabaseTransaction): Promise<Buffer[]> {
```
[the remark this answers is not in the document]
Comment (deekerno) [reply ironfish/src/indexers/minedBlocksIndexer.ts:286]: Oh yes, it should, good catch.

File: ironfish/src/indexers/database/sequenceToHashes.test.ts [hunk 35]
```diff
@@ -0,0 +1,23 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import {
+  SequenceToHashesValue,
+  SequenceToHashesValueEncoding,
+} from '../../blockchain/database/sequenceToHashes'
+
+describe('SequenceToHashesValueEncoding', () => {
+  describe('with a list of hashes', () => {
```
Comment (rohanjadvani) [inline ironfish/src/indexers/database/sequenceToHashes.test.ts:10]: Same comment here.

File: ironfish/src/indexers/minedBlocksIndexer.ts [hunk 36]
```diff
@@ -0,0 +1,383 @@
...
+
+          await this.updateHeadHash(header.hash, tx)
+        })
+      }
+    })
+
+    this.chainProcessor.onRemove.on(async (header) => {
+      await this.database.transaction(async (tx) => {
+        if (await this.minedBlocks.has(header.hash, tx)) {
+          const block = await this.chain.getBlock(header)
```
Comment (rohanjadvani) [inline ironfish/src/indexers/minedBlocksIndexer.ts:148]:
```suggestion
          const block = await this.chain.getBlock(header, tx)
```

**Review by rohanjadvani — APPROVED** [review]
